Datepicker - add ability to update the reference start/end dates

lyzhan7 opened the pull request:
### Platforms Impacted
- [x] iOS
- [ ] macOS
- [ ] win32 (Office)
- [ ] windows
- [ ] android

### Description of changes

Ask from a client team to be able to set the reference start/end dates (range of possible dates the calendar allows, not to be confused with "date range" which refers to the dates a user can select). 

The default reference start date is Jan 1, 3 years before the current year. The default reference end date is 10 years after the default reference start date. Ex. For May 23, 2022, the default range is Jan 1, 2019 - Dec 21, 2029.

This change allows the default Calendar Configuration to be overridden - which includes the reference start and end dates, as well as firstWeekday (usually begins on Sunday or Monday). Note that overriding the default calendar configuration affects **all** FURN datepickers. Additional work will be required if we want to support different calendar configurations for different datepickers.

### Verification

Added a button at the bottom of the FluentTester DatePicker test page that updates the default reference start/end dates to be from 1999 to 2000.

| Before | After |
|----------------------------------------------|--------------------------------------------|
| ![image](https://user-images.githubusercontent.com/78454019/169923134-4bc8469e-2b98-48dc-9595-42ce730fd8d3.png) | ![image](https://user-images.githubusercontent.com/78454019/169923462-d6d96d36-0931-4186-8230-0a07d1e901e2.png) | ![image](https://user-images.githubusercontent.com/78454019/169923134-4bc8469e-2b98-48dc-9595-42ce730fd8d3.png)
 

| Before (earliest possible start date you could scroll to) | After (earliest possible start date you could scroll to) |
|----------------------------------------------|--------------------------------------------|
| ![image](https://user-images.githubusercontent.com/78454019/169923204-43860ae9-bf22-46ed-82aa-4dd75d33a5b1.png) | ![image](https://user-images.githubusercontent.com/78454019/169923219-51e41698-0804-4d80-9602-6134cc630aaa.png) |

### Pull request checklist

This PR has considered (when applicable):
- [ ] Automated Tests
- [ ] Documentation and examples
- [ ] Keyboard Accessibility
- [ ] Voiceover
- [ ] Internationalization and Right-to-left Layouts


Notes:
- Not sure if using a button in the FluentTester datepicker page to change the default reference start end dates is best. Maybe a toggle would be better?
- This allows users to set the default "CalendarConfiguration", but this PR only exposes referenceStartDate and referenceEndDate. Do we also want to expose firstWeekday?
- With this change it is easier to get into a state where the selected date is outside the reference start/end dates, ex if the default start date is already out of the range, and you try to select the default date. There is an assert in datetimepickerviewdatasource (line 308, assert(dateIndex >= 0 && dateIndex < numberOfDates, "index out of bound")) that gets hit when this happens. Do we want to do anything about this?

lyzhan7 commented on apps/fluent-tester/src/FluentTester/TestComponents/NativeDatePicker/NativeDatePickerTest.tsx:
```diff
@@ -36,14 +38,14 @@ const NativeDatePickerMainTest: React.FunctionComponent = () => {
   return (
     <Stack style={stackStyle}>
-      <Text variant="headerStandard">Start Date/Time</Text>
+      <Text variant="headerStandard">Selected Start Date/Time</Text>
       <Text variant="subheaderStandard">
         {startDate?.toString()} {'\n'}
       </Text>
 
-      <Text variant="headerStandard">End Date/Time</Text>
+      <Text variant="headerStandard">Selected End Date/Time</Text>
       <Text variant="subheaderStandard">
-        {endDate?.toString()} {'\n'}
+        {endDate?.toString() ?? 'N/A'} {'\n'}
```
These changes aren't really related to being able to update the reference start/end dates - I made the labels more clear since "start date" is kind of ambiguous and also it was strange that if there was no end date, no placeholder text would show up. I can move this into another PR too

Saadnajmi commented:
Awesome, This looks great!  
A few notes:

- We generally don't mention specific partners or link internal work items in open source PRs. More often, we say "A client team asked..." =)
- The <Switch> control would work well as a toggle. We use it elsewhere.
- Might as well add it? Since it's the only other property you could configure
- I'm inclined to say it's ok to have that error, since that feels like a "The API was misused" error. If we don't like it, we could try catching the assert but I'm ok with leaving it as is.

lyzhan7 commented:
@Saadnajmi Thank you for your notes - updated the description, and decided a checkbox made more sense than a toggle since there was a visible difference when it was on or off. Added firstWeekday to it as well.

Updated screenshot:
![image](https://user-images.githubusercontent.com/78454019/170129257-c9ae1aaf-1bfd-4355-bc92-71228b4359bf.png)

Saadnajmi commented:
> @Saadnajmi Thank you for your notes - updated the description, and decided a checkbox made more sense than a toggle since there was a visible difference when it was on or off. Added firstWeekday to it as well.
> 
> Updated screenshot: ![image](https://user-images.githubusercontent.com/78454019/170129257-c9ae1aaf-1bfd-4355-bc92-71228b4359bf.png)

I think the iOS designs for Checkbox are not very finalized (iirc, the chevron is like super tiny?) and also I think the Apple HIG _technically_ wants to use Switches over checkboxes in most cases. This is me being picky, but would you actually change that to a Switch? I know @harrieshin mentioned before we shouldn't really use Checkbox on iOS much at all.

lyzhan7 commented:
Updated screenshot - using switch, removed firstWeekday

<img width="315" alt="Screen Shot 2022-05-24 at 4 09 40 PM" src="https://user-images.githubusercontent.com/78454019/170146340-4cc6b7df-1a5b-4842-8227-68a98ce5fff6.png">
